### ISSUE TITLE
[Customer] Using chvm -a to add additional disks does not preserve storageformat configuraiton

### ISSUE THREAD

**Issue by whowutwut** [xcat2/xcat-core#1557]
Customer brings up this issue using the xCAT 2.9.X release.   I'm opening to milestone in the 2.9 release to track it.  I think if we fix in master, we should roll this fix back to 2.9 or provide a patch to the customer. 

A vm node definition in xCAT is changed to use `vmstorageformat=raw` and re-created. The VM storage file is created correctly with type=raw.  However, when adding additional disks with the `chvm -a <size>` command, the additional disks created appear with the default `qcow2` format. 

Example:
```
[root@fs2vm100 ~]# lsdef victorvm1 -i vmstorageformat
Object name: victorvm1
    vmstorageformat=raw

[root@fs2 ~]# ls -ltr /var/lib/libvirt/images/victor*
-rw-------. 1 root root 10737418240 Jul 16 19:57 /var/lib/libvirt/images/victorvm1.sda.raw
```

Then creating additional disks
```
chvm victorvm1 -a 20G
chvm victorvm1 -a 10G
```

Results in the following files: 
```
[root@fs2 ~]# ls -ltr /var/lib/libvirt/images/victor*
-rw-------. 1 root root 10737418240 Jul 16 19:57 /var/lib/libvirt/images/victorvm1.sda.raw
-rw-------. 1 root root      197120 Jul 16 19:59 /var/lib/libvirt/images/victorvm1.sdb.qcow2 <---qcow2
-rw-------. 1 root root      197120 Jul 19 09:08 /var/lib/libvirt/images/victorvm1.sdc.qcow2 <--- qcow2
```

**Comment by daniceexi** [comment]
I think this is a bug of xCAT that we don't handle the storage format in the chvm command. Refer to the following code in the kvm.pm. We really should pass in the storage format when call the get_filepath_by_url in the chvm() subroutine.

~~~~
1925             #ok, now I need a volume created to attach
1926             push @diskstoadd, get_filepath_by_url(url => $location, dev => $dev, create => $_);
~~~~

**Comment by gurevichmark** [comment]
@daniceexi Thank you for pointing out the lines to in question. Before I implement the fix, there are 3 questions:

1. When creating vm with `mkvm` command, we first set the storage format to the value of `vmvirtflags imageformat=<value>` if specified. Then overwrite it with the value of `vmstorageformat=<value>` if specified. I assume `chvm` should do the same to figure out the value for the storage format ?

2. Currently the code for `chvm` does not check if vm is diskless before processing the `-a size` option, which results in the following hang:
>[root@fs2vm100 ~]# chvm mgXgeneratedvm1 -a 10
Error: kvm plugin bug, pid 2724, process description: 'xcatd SSL: chvm to mgXgeneratedvm1 for root@localhost: kvm instance' with error 'libvirt error code: 27, message: XML error: missing pool source name element
' while trying to fulfill request for the following nodes: mgXgeneratedvm1
 
   I think kvm.pm should display an error and exit in this case.

2. The man page for `chvm` in section for `VMware/KVM` shows `--resize <disk>=<size>` as a possible option to use. However looking at the code in kvm.pm, this option is accepted but then never used. As a result the following can be performed:
>[root@fs2vm100 ~]# chvm mgXgeneratedvm1 --resize junk=morejunk
[root@fs2vm100 ~]#

   I think we should remove this option from the `chvm` man page and from kvm.pm. Could it be that this option is valid only for VMware? I see some processing of it in esx.pm.

**Comment by daniceexi** [comment]
@gurevichmark 

1. Yes, ``chvm`` should use the same code logic as ``mkvm`` to determine the disk type;
2. We should not use the 'osimage' type to determine whether the vm can accept a disk. VM likes a baremetal server that even with a disk, the vm still can boot as diskless. We need figure out a better way to handle the ``-a`` for chvm;
3. ``chvm -a`` is used to add a new disk. ``--resize`` is used to resize a disk. So currently there's no any code for the flag in kvm.pm? If yes, we may consider remove this flag from man page.

**Comment by gurevichmark** [comment]
1. Will implement.
2. As discussed on Wednesday conference call, if `vmstorage` attribute is not defined, `chvm -a` will return an error.
3. It appears kvm does support volume resize with `vol-resize <vol> <capacity> [--pool <string>] [--allocate] [--delta] [--shrink]`. I will open a separate issue to fix that.

**Comment by gurevichmark** [comment]
This issue is fixed by pull request #1593 for xCAT 2.12 and pull request #1594 for xCAT 2.9.

**Comment by whowutwut** [comment]
Customer received the fix and said it resolved the issue.  Will close based on that comment.
